programs/storage_api/src/storage_processor.rs

```diff
@@ -45,120 +118,200 @@ pub fn process_instruction(
 
     if let Ok(syscall) = bincode::deserialize(data) {
         let mut storage_contract =
-            if let Ok(storage_contract) = bincode::deserialize(&keyed_accounts[0].account.data) {
-                storage_contract
-            } else {
-                StorageContract::default()
-            };
-
-        debug!(
-            "deserialized contract height: {}",
-            storage_contract.entry_height
-        );
+            read_contract_with_default(&syscall, &keyed_accounts[0].account)?;
         match syscall {
             StorageInstruction::SubmitMiningProof {
                 sha_state,
                 entry_height,
                 signature,
             } => {
-                let segment_index = get_segment_from_entry(entry_height);
-                let current_segment_index = get_segment_from_entry(storage_contract.entry_height);
-                if segment_index >= current_segment_index {
-                    return Err(InstructionError::InvalidArgument);
-                }
+                if let StorageContract::ReplicatorStorage { proofs, .. } = &mut storage_contract {
+                    if keyed_accounts.len() != 1 {
+                        // keyed_accounts[1] should be the main storage key
+                        // to access its data
+                        Err(InstructionError::InvalidArgument)?;
+                    }
 
-                debug!(
-                    "Mining proof submitted with contract {:?} entry_height: {}",
-                    sha_state, entry_height
-                );
-
-                let proof_info = ProofInfo {
-                    id: *keyed_accounts[0].signer_key().unwrap(),
-                    sha_state,
-                    signature,
-                };
-                storage_contract.proofs[segment_index].push(proof_info);
+                    let segment_index = get_segment_from_entry(entry_height);
```

Comment:
we need some kind of check to make sure this is in the current storage epoch, but maybe you are planning on adding that later?

Reply:
What determines a valid mining segment ? The verification should occur when a validator looks at these proofs.